test/metabase/query_processor/middleware/wrap_value_literals_test.clj

```diff
@@ -334,3 +334,20 @@
             preprocessed (qp.preprocess/preprocess query)]
         ;; [:query :filter 2 2 :database_type] points to wrapped value's options
         (is (= "BIGINT" (get-in preprocessed [:query :filter 2 2 :database_type])))))))
+
+(deftest ^:parallel type-info-gets-field-ref-data-test
+  (testing "type info includes data from :field_ref as well as :source-metadata"
+    (is (=? {:query
+             {:filter
+              [:=
+               [:field "CATEGORY" {:base-type :type/Text}]
+               [:value "Doohickey" {:base_type :type/Text,
+                                    :effective_type :type/Text,
+                                    :database_type "CHARACTER VARYING"}]]}}
```

Comment:
Please wrap this with `mt/mbql-query` too; it does more than fill in field IDs. It does some `=?` magic to ignore the drift of randomized values.